File: scribbleshare-commons/src/main/java/net/stzups/scribbleshare/data/objects/canvas/CanvasObjectWrapper.java
package net.stzups.scribbleshare.data.objects.canvas;

import io.netty.buffer.ByteBuf;
import net.stzups.scribbleshare.data.objects.canvas.canvasObject.CanvasObject;
import net.stzups.scribbleshare.data.objects.canvas.canvasObject.CanvasObjectType;
import net.stzups.scribbleshare.data.objects.exceptions.DeserializationException;

public class CanvasObjectWrapper {
    private final CanvasObjectType type;
    private final CanvasObject canvasObject;

    public CanvasObjectWrapper(CanvasObjectType type, CanvasObject canvasObject) {
        this.type = type;
        this.canvasObject = canvasObject;
    }

    public CanvasObjectWrapper(ByteBuf byteBuf) throws DeserializationException {
        this.type = CanvasObjectType.deserialize(byteBuf.readUnsignedByte());
        this.canvasObject = CanvasObject.deserialize(type, byteBuf);
    }

    public void serialize(ByteBuf byteBuf) {
        byteBuf.writeByte((byte) type.getId());
        canvasObject.serialize(byteBuf);
    }

    CanvasObjectType getType() {
        return type;
    }

    public CanvasObject getCanvasObject() {
        return canvasObject;
    }
}

File: scribbleshare-app/scribbleshare-frontend/src/scripts/canvas/canvasObject/CanvasObjectWrapper.js
import {getCanvasObject} from "./getCanvasObject.js";

export default class CanvasObjectWrapper {
    constructor(canvasObjectType, canvasObject) {
        this.canvasObjectType = canvasObjectType;
        this.canvasObject = canvasObject;
    }

    serialize(writer) {
        writer.writeUint8(this.canvasObjectType);
        this.canvasObject.serialize(writer);
    }

    static deserialize(reader) {
        let canvasObjectWrapper = Object.create(this.prototype);
        canvasObjectWrapper.canvasObjectType = reader.readUint8();
        canvasObjectWrapper.canvasObject = getCanvasObject(canvasObjectWrapper.canvasObjectType, reader);
        return canvasObjectWrapper;
    }
}